include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -4418,6 +4421,666 @@ namespace internal
+                                       c_mask,
+                                       values);
+
+      return false;
+    }
+
+  private:
+    template <int fe_degree, unsigned int side, bool transpose>
+    static void
+    interpolate_2D(const unsigned int given_degree,
+                   bool               is_subface_0,
+                   const unsigned int v,
+                   const Number *     weight,
+                   Number *           values)
+    {
+      if (is_subface_0)
+        interpolate_2D<fe_degree, side, transpose, true>(given_degree,
+                                                         v,
+                                                         weight,
+                                                         values);
+      else
+        interpolate_2D<fe_degree, side, transpose, false>(given_degree,
+                                                          v,
+                                                          weight,
+                                                          values);
+    }
+
+    template <int          fe_degree,
+              unsigned int side,
+              bool         transpose,
+              bool         is_subface_0>
+    static void
+    interpolate_2D(const unsigned int given_degree,
+                   const unsigned int v,
+                   const Number *     weight,
+                   Number *           values)
+    {
+      typename Number::value_type temp[10 /*TODO*/];
+
+      const unsigned int points =
+        (fe_degree != -1 ? fe_degree : given_degree) + 1;
+
+      const unsigned int d = side / 2; // direction
+      const unsigned int s = side % 2; // left or right surface
+
+      const unsigned int offset = dealii::Utilities::pow(points, d + 1);
+      const unsigned int stride =
+        (s == 0 ? 0 : (points - 1)) * dealii::Utilities::pow(points, d);
+
+      const unsigned int r1 = dealii::Utilities::pow(points, dim - d - 1);
+      const unsigned int r2 = dealii::Utilities::pow(points, d);
+
+      // copy result back
+      for (unsigned int i = 0, k = 0; i < r1; ++i)
+        for (unsigned int j = 0; j < r2; ++j, ++k)
+          temp[k] = values[i * offset + stride + j][v];
+
+      // perform interpolation point by point (note: r1 * r2 == points^(dim-1))
+      for (unsigned int i = 0, k = 0; i < r1; ++i)
+        for (unsigned int j = 0; j < r2; ++j, ++k)
+          for (unsigned int h = 0; h < points; ++h)
+            {
+              const unsigned int index_v = i * offset + stride + j;
+              const unsigned int index_w =
+                ((transpose ? 1 : points) *
+                 (is_subface_0 ? k : (points - 1 - k))) +
+                ((transpose ? points : 1) *
+                 (is_subface_0 ? h : (points - 1 - h)));
+              if (h == 0)
+                values[index_v][v] = weight[index_w][v] * temp[h];
+              else
+                values[index_v][v] += weight[index_w][v] * temp[h];
```

Comment:
Did you check the assembly code of this part? The reason I'm asking is that I believe the compiler will generate a store operation on every iteration because it won't know that `weight` and `values` do not alias. If the compiler is smart, which I believe it is, it will at least not re-load the entry every time, so it is only the stores here. The other tensor-product functions use a local variable `Number sum;` that is incremented inise the loop over `h` and written once into the result array at the end of the loop for this purpose.